src/lua/uri.lua

```diff
@@ -169,8 +178,63 @@ local function format(uri, write_password)
     return str
 end
 
+--- Encodes the given string
+-- Uses Curl function curl_easy_escape(),
+-- see description in https://curl.haxx.se/libcurl/c/curl_easy_escape.html
+-- @function url_encode
+-- @string buf - string to be encode
+-- @returns result string or nil, err
+local function url_encode(buf)
+    if type(buf) ~= 'string' then
+        error("Usage: uri.encode(string)")
+    end
+
+    if not curl then
+        return nil, 'curl handle is nil'
+    end
+
+    local escaped_str = builtin.curl_easy_escape(curl, buf, #buf)
+    if escaped_str == nil then
+        return nil, 'curl_easy_escape() error'
+    end
+
+    local res = ffi.string(escaped_str)
+    builtin.curl_free(escaped_str)
+
+    return res
+end
+
+--- Decodes the given string
+-- Uses Curl function curl_easy_unescape(),
+-- see description in https://curl.haxx.se/libcurl/c/curl_easy_unescape.html
+-- @function url_decode
+-- @string buf - encoded URL
+-- @returns string or nil, err
+local function url_decode(buf)
+    if type(buf) ~= 'string' then
+        error("Usage: uri.decode(string)" .. type(buf))
+    end
+
+    if not curl then
+        return nil, 'curl handle is nil'
+    end
+
+    local url_decode_outlength = ffi.new("int[1]")
+    local unescaped_str = builtin.curl_easy_unescape(curl, buf, #buf, url_decode_outlength)
+    if unescaped_str == nil then
+        return nil, 'curl_easy_unescape() error'
+    end
+
+    local res = ffi.string(unescaped_str, url_decode_outlength[0])
```

Comment:
It should be tested. String that contains zero byte could be stripped here if url_decode_outlength[0] won't provided. Currently I don't see such test.